### ISSUE TITLE
"Spanned" misalignment with non-symmetrical display arrangements

### ISSUE THREAD

**Issue by theoriginalbit** [JohnCoates/Aerial#835]
**Hardware:** MacBook Pro (15-inch, 2018), macOS 10.14.6
**Displays:**
<img width="698" alt="image" src="https://user-images.githubusercontent.com/1377564/62024652-29662f80-b219-11e9-984a-a064853436bb.png">
<img width="780" alt="image" src="https://user-images.githubusercontent.com/1377564/62024680-3e42c300-b219-11e9-9393-6c0cb1b53502.png">

#### Description of issue / Feature request

Adding a horizontal margin to the screens results in my built-in (bottom row) display being misaligned by half of the margin to the left.

<img width="780" alt="image" src="https://user-images.githubusercontent.com/1377564/62024747-70542500-b219-11e9-9f38-61021ed65c60.png">

**Comment by glouel** [comment]
Hi @theoriginalbit 

For this kind of configurations, I'm afraid the basic margin support I gave is insufficient, as while I could probably compute to add half the margin for your specific case, the algorithm gets crazy complex with more screens. 

For those more complex configurations, some sort of a manual mode where you can edit individual margins is probably better, I'll put that on my todo list.

**Comment by theoriginalbit** [comment]
Thanks for the quick response @glouel 

I knew that the algorithm would definitely start to get crazy as you'd have to detect the original arrangement and try to infer the differing margins. So I wanted to make sure that these kinds of (strange) arrangements are on the radar as something that could be supported. Manual margins editing sounds like the solution.

**Comment by duecedriver** [comment]
to get this to work right, with different displays of different dot pitch.. it DPI you would need to have an algorithm similar to photoshop that lets you tell the aerial screensaver algorithms the DPI of each monitor, thereby allowing the algorithm to draw out each image in a size appropriate fashion

screen resolution and position as identified by OS X is not enough information as you can have 2 identically physically sized monitors sitting side by side.. it say 27" diagonal .. but one is 1080 and the other is 4k.. obviously if the algo can't just split the image down the middle and have it meet up

the simplest way for the user interface would be for the screensaver to ask for the diagonal measurement be input by the user.. then the screensaver algo, using this diagonal size and operating system reported orientation,position,and resolution should be able to come up with a scaling factor

**Comment by glouel** [comment]
macOS does handle this on its own, by separating pixel size and points, using a scale factor. 

Each screen has a size in points where the point is about the same physical size. That point may be represented by one or more pixels to handle the difference. Example, you have a 27" 5K iMac and a 27" 2560x1440 external display, both will have the same point size, but one will have a scaleFactor of 2, the other of 1. Scale factors don't have to be integer, they are notoriously not on 15" MBP for example by default. 

macOS handles this gracefully, all coordinates that macOS programs uses are in point and not pixels and that solves most of the issues. So Aerial already does this and handles correctly those situations.

**Comment by duecedriver** [comment]
yeah.. and how is that handled on a mac pro with external monitors apple knows nothing about?

**Comment by glouel** [comment]
By using EDID to read each monitor's DPI ? There's no magic here, macOS knows the display's physical size in millimeters. 

https://en.wikipedia.org/wiki/Extended_Display_Identification_Data

This has nothing to do with the algorithm issue mentioned discussed in this issue in any case, or are you saying you are seeing a scaling bug on your configuration ? If so please detail it in a new issue, as this one regards a specific bug.

**Comment by duecedriver** [comment]
There are a host of monitors and a plethora of hdtvs that don’t Inject their edid correctly

And there are numerous instances like Vega GPUs that are not reading edid or getting edid values and that is giving users a host of issues like black screen from sleep wake etc

I would not code to expect clean edid info

**Comment by glouel** [comment]
If someone, someday, reports an issue, I'll be sure to look at it...

**Comment by duecedriver** [comment]
I am pretty sure that people already are. Hence the black screen and login windows issue. If the screensaver is calling for edid and not getting it that could break the handshake connection of the video signal especially DisplayPort

**Comment by glouel** [comment]
I've already replied to you in the other issue, regarding your crash with version 1.5.0, and still waiting for an answer from you there on your macOS version and if you tried the latest beta I linked to you, twice, that has a specific fix for multi monitor setups over changes introduced in 1.5.0. Considering 1.4.9 works correctly for you as you mentioned yourself, EDID is probably not behind your crash since that part of the screen saver didn't move. 

FYI, the issue you replied into was regarding a completely different problem with 3rd party firewall Little Snitch and the auto update mechanism in Aerial. Still, I'm trying to help you here but I'd appreciate a tiny bit of cooperation from you if you want help from me. Thanks.

**Comment by duecedriver** [comment]
Already stated in my reply. Mojave 10.14.6

And beta didn’t fix

**Comment by glouel** [comment]
> Already stated in my reply. Mojave 10.14.6

I don't see where but ok. 

> And beta didn’t fix

Ok.

Best way forward, I'm gonna need your crash log. Please check this post (the top part) about where to find it and what it's called. That should tell us what exactly crashed for you : https://github.com/JohnCoates/Aerial/issues/681#issuecomment-443729062

**Comment by duecedriver** [comment]
ok... I deleted most of what I could find running mojave by the wayre-installed 1.4.9 I think.. last non beta prior to 1.5selected h264 and tested.. workedupgraded to 1.5 againsame thing jammed upkillall loginwindow from remote machine gets me back inwent in selected single screen and h264.. workedselected single screen h265, workedselected 4k hevc.. now working.. but not reliablysome hook between the screensaver and loginwindow process is likely not syncing up
--

**Comment by duecedriver** [comment]
there are no crash logs... 


it is either freezing loginwindow or perhaps its waiting on something that never arrives but I see nothing in the logs, I just ssh in from another machine and killall longinwindow to kill your screensaver black screen login lockup and then I am back in .. system stays running

**Comment by glouel** [comment]
Ok. I would suggest this then:
- Install 1.5.1beta11
- In advanced, enable "debug mode" and "log to disk"
- Reproduce the issue
- Paste the log (you can access Aerial.log with the Show in Finder button)

I would suggest disabling Auto updates too just in case as this has been known to cause issues with some 3rd party software. Would be helpful to know which Mac model you have too, if your problem only happen with 4K HEVC if I understand what you say.

**Comment by petrsimanek** [comment]
Hi, it seems to me that the 1.5.1beta11 solved the problem for me. Thank you.

**Comment by theoriginalbit** [comment]
> solved the problem for me

The problem I raised in this issue, or the hijacked problem?

**Comment by petrsimanek** [comment]
I meant the problem you raised in this issue (the misalignments in displays)

**Comment by glouel** [comment]
@theoriginalbit & @petrsimanek 

I think beta11 solves an issue when you have two identical screens. Previous builds would sometimes mix them up and that probably had an effect on margins. With that said, for the exact setup of @theoriginalbit I don't think I solved it yet but for other setups it might look like it.

**Comment by glouel** [comment]
So @theoriginalbit, sorry it took me a while, but I finally got to implement it. 

<img width="799" alt="Capture d’écran 2019-09-11 à 17 02 53" src="https://user-images.githubusercontent.com/37544189/64713060-8b72af80-d4bc-11e9-9226-87694895168a.png">

This is not the prettiest thing but it should do what you want. There's a new checkbox for advanced mode, before checking it, set some values you'd want to work in the margin fields, as it will help a bit. Then press "Edit", you'll see a generated JSON that lists your displays. 

Each screen looks like this : 

```
    {
      "ztop" : -1440,
      "zleft" : 0,
      "offsettop" : 0,
      "offsetleft" : 10
    }
```

ztop/zleft, don't touch, that's the position of your screen in the span of all your monitors (the spans origin is at the bottom left, which is a bit counter intuitive, that's a macOS thing).

What you change is the offsettop/offsetleft, which are the horizontal and vertical offsets. They can  be negative too, and you can create crazy arrangements just like my screenshots where screens overlap. Press "Apply" to see the preview updates itself without closing the editor. What you see there is what you'll get in the real screensaver.

This is in beta12 : 
https://github.com/JohnCoates/Aerial/releases/download/v1.5.1beta12/Aerial.saver.zip

Please let me know if that works correctly.

**Comment by glouel** [comment]
Closing with the release of 1.6.0 finale.
